tensorboard/webapp/widgets/line_chart_v2/lib/series_line_view.ts

```diff
@@ -15,19 +15,140 @@ limitations under the License.
 
 import {DataDrawable} from './drawable';
 
+enum PartitionType {
+  NUMBER,
+  NAN,
+}
+
 export class SeriesLineView extends DataDrawable {
+  private partitionPolyline(
+    polyline: Float32Array
+  ): Array<{polyline: Float32Array; type: PartitionType}> {
+    if (polyline.length % 2 !== 0) {
+      throw new Error(`Cannot have odd length-ed polyline: ${polyline.length}`);
+    }
+
+    const partition = [];
+    let partitionStartInd: number = 0;
+    let isPrevValueNaN = false;
+    const zeroCoord = this.coordinator.transformDataToUiCoord(
+      this.getLayoutRect(),
+      [0, 0]
+    );
+    const zeroPoint = {x: zeroCoord[0], y: zeroCoord[1]};
+
+    let lastLegalNumber: {
+      x: number;
+      y: number;
+    } | null = null;
+
+    function recordPartition(
+      isNumberPartition: boolean,
+      slice: Float32Array,
+      nanSubstitude: {x: number; y: number}
+    ) {
+      if (isNumberPartition) {
+        return {type: PartitionType.NUMBER, polyline: slice};
+      } else {
+        return {
+          type: PartitionType.NAN,
+          polyline: slice.map((x, ind) => {
+            if (!isNaN(x)) return x;
+            return ind % 2 === 0 ? nanSubstitude.x : nanSubstitude.y;
+          }),
+        };
+      }
+    }
+
+    for (let index = 0; index < polyline.length; index += 2) {
+      const x = polyline[index];
+      const y = polyline[index + 1];
+      const hasNaN = isNaN(x) || isNaN(y);
+      if (hasNaN !== isPrevValueNaN && partitionStartInd !== index) {
+        partition.push(
+          recordPartition(
+            !isPrevValueNaN,
+            polyline.slice(partitionStartInd, index),
+            lastLegalNumber === null ? {x, y} : lastLegalNumber
+          )
+        );
+        partitionStartInd = index;
+      }
+
+      if (!hasNaN) {
+        lastLegalNumber = {x, y};
+      }
+
+      isPrevValueNaN = hasNaN;
+    }
+
+    if (partitionStartInd !== polyline.length - 1) {
+      partition.push(
+        recordPartition(
+          !isPrevValueNaN,
+          polyline.slice(partitionStartInd, polyline.length),
+          lastLegalNumber ?? zeroPoint
+        )
+      );
+    }
+
+    return partition;
+  }
+
   redraw() {
     for (const series of this.series) {
       const map = this.getMetadataMap();
       const metadata = map[series.id];
       if (!metadata) continue;
 
-      this.paintBrush.setLine(series.id, series.polyline, {
-        color: metadata.color,
-        visible: metadata.visible || false,
-        opacity: metadata.opacity ?? 1,
-        width: 1,
-      });
+      const partitionedPolyline = this.partitionPolyline(series.polyline);
+
+      for (const [
+        partitionInd,
+        {type, polyline},
+      ] of partitionedPolyline.entries()) {
+        if (type === PartitionType.NUMBER) {
+          if (polyline.length === 2) {
+            if (metadata.aux) continue;
```

Comment:
I agree that it has leaky abstraction and ideally, we should make NaN drawing a feature specified in the metadata but I do not yet see a strong reason to do so.

Few things though (since there seems to be some confusion).
- aux is the original series (shown in faded color) when smoothing is enabled
- vz_line_chart did not show NaNs for aux/original-series and I am merely trying to have the same UX.

vz_line_chart
![image](https://user-images.githubusercontent.com/2547313/102136058-bbbaa380-3e0d-11eb-99bd-bba22fa2c92d.png)

Gpu line chart
![image](https://user-images.githubusercontent.com/2547313/102136123-d1c86400-3e0d-11eb-87de-7641a3f189c2.png)

Again, we are kind of are abusing the `aux` flag but I think there is a balance between flags for every feature vs. simplicity. I expect the API to naturally evolve as we adopt the line chart in more use cases.

Reply:
Hm, I think I see where you are coming from. By making SeriesLineView aware of the fact that aux lines never draw NaNs, we gain a good amount of simplicity.  I think this is reasonable, and keeping L135-136 is fine with me.

However, I don't agree in the number case (non-NaN).  Shouldn't all (non-NaN) lines be drawn the same way, whether they are aux or not?  If we treat aux+nonNaN the same as nonAux+nonNan, I think we satisfy those goals:

> aux is the original series (shown in faded color) when smoothing is enabled
> vz_line_chart did not show NaNs for aux/original-series and I am merely trying to have the same UX.

Concretely, I think dropping L111 changes our cases from this:
If line segment is [number && singlePoint && aux] skip
If line segment is [number && non-aux] draw
If line segment is [NaN && aux] skip
If line segment is [NaN && non-aux] draw

Into this:
If line segment is [number] draw
If line segment is [NaN && aux] skip
If line segment is [NaN && non-aux] draw

Is there something we gain by keeping L111, besides rendering fewer points in the single point scenario, that I might be missing?

Reply:
Ah, got it. I simply decided not to draw the circle for aux because there was no circle for single point in vz_line_chart at all! 

I took your advise and remove L111.